### PR TITLE
async member update optimization

### PR DESCRIPTION
**Pull request by JonasSteur**
The original approach was to give the responsibility of handling the member update batches to mailchimp by using batch operations ( http://developer.mailchimp.com/documentation/mailchimp/reference/batches/)
In practice this turns out to be quite unreliable when doing regular big updates as the way mailchimp processes the batch operation queue doesn't seem to be FIFO. This is a pain in the ass to say the least.

The new approach is to do direct calls to the /lists endpoint with batches of 500 members. I tested this and I succeeded in adding/updating 400 000 members under an hour this way. 

One of the challenges now will be handling the 10 connections limit ( as there are other scheduled budskap tasks that could use a connection as well during the time this update is running)

For progress tracking a callback option has been added. The callback should be a generator like this:
```python
def callback_function():
    while True:
      progress = yield
      <do_something_with>(progress)
```

### CODE REVIEW

File: wingmonkey/async_operations.py [hunk 1]
```diff
@@ -1,21 +1,48 @@
...
+        if not isinstance(callback, GeneratorType):
+            raise TypeError(f'callback should be {GeneratorType} but got {type(callback)} instead')
+
+        self.callback = callback
+        self.id = uuid4()
+        self.total = total
+        self.completed = 0
+        self.last_response_status = None
+
+        next(callback)
```
Comment (mhemeryck) [inline wingmonkey/async_operations.py]: Won't this already do the first iteration in the initialization of the object? (might be intended behavior though).

File: wingmonkey/async_operations.py [hunk 2]
```diff
@@ -42,23 +69,120 @@
...
     :param results: list
+    :param json: Boolean: Return response json instead of whole instance
+    :param status_only: Boolean: Only return response status instead of whole instance
+    :param progress: Progress instance
     """
     while not queue.empty():
         task = await queue.get()
-        results.append(await _async_task(**task))
+        batch_size = task.get('batch_size', 0)
+        task.pop('batch_size', None)
```
Comment (mhemeryck) [inline wingmonkey/async_operations.py]: Doesn't this `pop` command already return the `batch_size`? I.e. can't you do this in a single line?

File: wingmonkey/async_operations.py [hunk 3]
```diff
@@ -42,23 +69,120 @@
...
+        response = await _async_task(**task)
+
+        if status_only or isinstance(response, ClientException):
+            result = response.status
+        elif json:
+            result = await response.json()
+        else:
+            result = response
```
Comment (mhemeryck) [inline wingmonkey/async_operations.py]: Is it possible / desired to have a results list with different result types? Seems you have a single function here that handles a different number of possible implementations / side effects.

File: wingmonkey/async_operations.py (hunk 1 above)
Comment (JonasSteur) [reply wingmonkey/async_operations.py]: This is intended behavior indeed to 'prime' the generator.

File: wingmonkey/async_operations.py (hunk 2 above)
Comment (JonasSteur) [reply wingmonkey/async_operations.py]: Yes, completely true. Let's fix this.

File: wingmonkey/async_operations.py (hunk 3 above)
Comment (JonasSteur) [reply wingmonkey/async_operations.py]: It became a bit of a frankenstein monster indeed. Initially it was just returning any result it got from the task but now it is actually doing something to the data before passing it on. The result can be a (aio) ClientResponse but also a ClientException when errors occur... Not really happy with the way it looks now. It works but it's quite messy. Will think about a better way to design this , probably putting the result parsing in a different function.

**Pull request merged**